### ISSUE TITLE
[Bug] KubeRay cluster fails to start whenever worker memory limits >=4GiB

### ISSUE THREAD

**Issue by jonded94** [ray-project/kuberay#2640]
### Search before asking

- [X] I searched the [issues](https://github.com/ray-project/kuberay/issues) and found no similar issues.

### KubeRay Component

ray-operator

### What happened + What you expected to happen

Hello,

I'm using KubeRay on an internal k8s cluster with really beefy machines (hundreds of cores, 2.3TiB RAM).

Setting up clusters in general through Helm Charts and using the KubeRay operator works without a problem:
```
helm install testcluster kuberay/ray-cluster -f values.yaml --version 1.2.2 
```

But, as soon as I raise the Ray worker memory limits to a value equal or above 4GiB, no worker is able anymore to connect to the head pod. This is entirely reproducable and there is an extremely sharp boundary between where it works and where it doesn't.

```
worker:
  resources:
    limits:
      cpu: "1"
      memory: "4Gi"  # <-- this setting is changed
    requests:
      cpu: "1"
      memory: "1G"
```

This is how it looks like for when `limits` is below 4GiB:
```
Defaulted container "ray-worker" out of: ray-worker, wait-gcs-ready (init)
[2024-12-12 04:31:44,938 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 6fdc2334df20c69d556553032fb4b03e7e9cfa0f53d9aba053669b63
2024-12-12 04:31:44,878	INFO scripts.py:946 -- Local node IP: 10.236.217.236
2024-12-12 04:31:45,940	SUCC scripts.py:959 -- --------------------
2024-12-12 04:31:45,941	SUCC scripts.py:960 -- Ray runtime started.
2024-12-12 04:31:45,941	SUCC scripts.py:961 -- --------------------
2024-12-12 04:31:45,941	INFO scripts.py:963 -- To terminate the Ray runtime, run
2024-12-12 04:31:45,941	INFO scripts.py:964 --   ray stop
2024-12-12 04:31:45,941	INFO scripts.py:972 -- --block
2024-12-12 04:31:45,941	INFO scripts.py:973 -- This command will now block forever until terminated by a signal.
2024-12-12 04:31:45,941	INFO scripts.py:976 -- Running subprocesses are monitored and a message will be printed if any of them terminate unexpectedly. Subprocesses exit with SIGTERM will be treated as graceful, thus NOT reported.
```

This is how it looks like for it's above or equal 4GiB:
```
Defaulted container "ray-worker" out of: ray-worker, wait-gcs-ready (init)
2024-12-12 04:56:06,843	WARNING services.py:2009 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 1000001536 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=1.28gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
[2024-12-12 04:56:06,933 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:07,934 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:08,935 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:09,936 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:10,937 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:11,938 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:12,939 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:13,940 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:14,941 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
[2024-12-12 04:56:15,941 W 8 8] global_state_accessor.cc:437: Retrying to get node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13
Traceback (most recent call last):
  File "/home/ray/anaconda3/bin/ray", line 8, in <module>
2024-12-12 04:56:06,823	INFO scripts.py:946 -- Local node IP: 10.236.219.78
    sys.exit(main())
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/scripts/scripts.py", line 2612, in main
    return cli()
  File "/home/ray/anaconda3/lib/python3.10/site-packages/click/core.py", line 1157, in __call__
    return self.main(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/click/core.py", line 1078, in main
    rv = self.invoke(ctx)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/click/core.py", line 1688, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/ray/anaconda3/lib/python3.10/site-packages/click/core.py", line 1434, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/click/core.py", line 783, in invoke
    return __callback(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/autoscaler/_private/cli_logger.py", line 856, in wrapper
    return f(*args, **kwargs)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/scripts/scripts.py", line 948, in start
    node = ray._private.node.Node(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/node.py", line 349, in __init__
    node_info = ray._private.services.get_node(
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/services.py", line 486, in get_node
    return global_state.get_node(node_id)
  File "/home/ray/anaconda3/lib/python3.10/site-packages/ray/_private/state.py", line 806, in get_node
    return self.global_state_accessor.get_node(node_id)
  File "python/ray/includes/global_state_accessor.pxi", line 270, in ray._raylet.GlobalStateAccessor.get_node
RuntimeError: b'GCS cannot find the node with node ID 09be14ce03b62ea760f96cce6619bc102a8c41ab9121ec92ce821c13'
```

### Reproduction script

```
image:
  repository: rayproject/ray
  tag: 2.40.0-py310
  pullPolicy: IfNotPresent

head:
  resources:
    limits:
      cpu: "1"
      # To avoid out-of-memory issues, never allocate less than 2G memory for the Ray head.
      memory: "10G"
    requests:
      cpu: "1"
      memory: "2G"
  podSecurityContext:
    fsGroup: 1000
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL
    privileged: false
    readOnlyRootFilesystem: true
    runAsNonRoot: true
    runAsUser: 1000
    seccompProfile:
      type: RuntimeDefault

worker:
  resources:
    limits:
      cpu: "1"
      memory: "5Gi"
    requests:
      cpu: "1"
      memory: "1G"
  podSecurityContext:
    fsGroup: 1000
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL
    privileged: false
    readOnlyRootFilesystem: true
    runAsNonRoot: true
    runAsUser: 1000
    seccompProfile:
      type: RuntimeDefault
```

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by jonded94** [comment]
I could isolate it even further: When setting memory requests to 2GiB, limits to 4GiB it works.
For memory limits set to 20GiB, the threshold for requests seems to be 7GiB.

So apparently there is some issue that occurs whenever the ratio of memory limits to request is greater than 2.5 (or sth. like that)?

Is this expected? I'm very puzzled about this behaviour.

**Comment by andrewsykim** [comment]
That is odd behavior, can you share the full YAML of the RayCluster from the server? (kubectl get raycluster ray-cluster -o yaml)

**Comment by jonded94** [comment]
```
$ kubectl get raycluster testcluster-4gib -o yaml
apiVersion: ray.io/v1
kind: RayCluster
metadata:
  annotations:
    meta.helm.sh/release-name: testcluster-4gib
    meta.helm.sh/release-namespace: dev-namespace
  creationTimestamp: "2024-12-12T14:54:49Z"
  generation: 1
  labels:
    app.kubernetes.io/instance: testcluster-4gib
    app.kubernetes.io/managed-by: Helm
    helm.sh/chart: ray-cluster-1.2.2
  name: testcluster-4gib-kuberay
  namespace: dev-namespace
  resourceVersion: "1804595063"
  uid: 0254451b-513c-41b9-957e-14cb43b2c97a
spec:
  headGroupSpec:
    rayStartParams:
      dashboard-host: 0.0.0.0
    serviceType: ClusterIP
    template:
      metadata:
        annotations: {}
        labels:
          app.kubernetes.io/instance: testcluster-4gib
          app.kubernetes.io/managed-by: Helm
          helm.sh/chart: ray-cluster-1.2.2
      spec:
        affinity: {}
        containers:
        - image: rayproject/ray:2.40.0-py310
          imagePullPolicy: IfNotPresent
          name: ray-head
          resources:
            limits:
              cpu: "1"
              memory: 10G
            requests:
              cpu: "1"
              memory: 2G
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
              - ALL
            privileged: false
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1000
            seccompProfile:
              type: RuntimeDefault
          volumeMounts:
          - mountPath: /tmp/ray
            name: log-volume
        imagePullSecrets: []
        nodeSelector: {}
        securityContext:
          fsGroup: 1000
        tolerations: []
        volumes:
        - emptyDir: {}
          name: log-volume
  workerGroupSpecs:
  - groupName: workergroup
    maxReplicas: 3
    minReplicas: 1
    numOfHosts: 1
    rayStartParams: {}
    replicas: 1
    template:
      metadata:
        annotations: {}
        labels:
          app.kubernetes.io/instance: testcluster-4gib
          app.kubernetes.io/managed-by: Helm
          helm.sh/chart: ray-cluster-1.2.2
      spec:
        affinity: {}
        containers:
        - image: rayproject/ray:2.40.0-py310
          imagePullPolicy: IfNotPresent
          name: ray-worker
          resources:
            limits:
              cpu: "1"
              memory: 4Gi
            requests:
              cpu: "1"
              memory: 1G
          securityContext:
            allowPrivilegeEscalation: false
            capabilities:
              drop:
              - ALL
            privileged: false
            readOnlyRootFilesystem: true
            runAsNonRoot: true
            runAsUser: 1000
            seccompProfile:
              type: RuntimeDefault
          volumeMounts:
          - mountPath: /tmp/ray
            name: log-volume
        imagePullSecrets: []
        nodeSelector: {}
        securityContext:
          fsGroup: 1000
        tolerations: []
        volumes:
        - emptyDir: {}
          name: log-volume
status:
  availableWorkerReplicas: 1
  desiredCPU: "2"
  desiredGPU: "0"
  desiredMemory: 3G
  desiredTPU: "0"
  desiredWorkerReplicas: 1
  endpoints:
    client: "10001"
    dashboard: "8265"
    metrics: "8080"
    redis: "6379"
    serve: "8000"
  head:
    podIP: 10.236.217.172
    podName: testcluster-4gib-kuberay-head-fzp9n
    serviceIP: 10.236.217.172
    serviceName: testcluster-4gib-kuberay-head-svc
  lastUpdateTime: "2024-12-12T14:55:37Z"
  maxWorkerReplicas: 3
  minWorkerReplicas: 1
  observedGeneration: 1
```

I think I found the reason though. The worker pods have an additional volume `shared-mem` that apparently has the same size as the memory `requests` (here is an example with 4GiB limits, 2GiB requests):
```
$ kubectl get pod testcluster-4gib-2gib-kuberay-workergroup-worker-4hvds -o yaml
[...]
  - emptyDir:
      medium: Memory
      sizeLimit: 2G
    name: shared-mem
```
I was made aware of that because of the warning message at the start of the logs:
```
ray-worker 2024-12-12 07:01:45,506    WARNING services.py:2022 -- WARNING: The object store is using /tmp instead of /dev/shm because /dev/shm has only 1000001536 bytes available. This will harm performance! You may be able to free up space by deleting files in /dev/shm. If you are inside a Docker container, you can increase /dev/shm size by passing '--shm-size=1.27gb' to 'docker run' (or add it to the run_options list in a Ray cluster config). Make sure to set this to more than 30% of available RAM.
```
Setting this manually to at least 30% of RAM, so for example to two 2GiB, will lead to a working cluster again. This is such an example worker config:
```
worker:
  replicas: 1
  minReplicas: 1
  maxReplicas: 3
  resources:
    limits:
      cpu: "1"
      memory: "4Gi"
    requests:
      cpu: "1"
      memory: "1G"
  podSecurityContext:
    fsGroup: 1000
  securityContext:
    allowPrivilegeEscalation: false
    capabilities:
      drop:
        - ALL
    privileged: false
    readOnlyRootFilesystem: true
    runAsNonRoot: true
    runAsUser: 1000
    seccompProfile:
      type: RuntimeDefault
  volumes:
    - name: shared-mem
      emptyDir:
        sizeLimit: 2Gi
    - name: log-volume
      emptyDir: {}
```

To be honest, I don't really understand ***why*** though. How are these things connected?

**Comment by jonded94** [comment]
Moreover, on a *not* working cluster, so where this `shared-mem` volume is set by default to the same capacity as memory requests, it has the additional field `medium: Memory`:

```
- emptyDir:
    medium: Memory
    sizeLimit: 1G 
  name: shared-mem               
```

**Comment by andrewsykim** [comment]
Can you share the full output of `kubectl get pod testcluster-4gib-2gib-kuberay-workergroup-worker-4hvds -o yaml`?

**Comment by jonded94** [comment]
Ok! This is the worker pod with 4GiB Memory Limits, 1GiB Memory Requests but a 2GiB `/dev/shm` volume:
```
$ kubectl get pod testcluster-4gib-1gib-2gib-volume-kuberay-workergr-worker-h7brl -o yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    kubernetes.io/limit-ranger: 'LimitRanger plugin set: ephemeral-storage request
      for container ray-worker; ephemeral-storage request for init container wait-gcs-ready'
    ray.io/ft-enabled: "false"
  creationTimestamp: "2024-12-12T16:04:28Z"
  generateName: testcluster-4gib-1gib-2gib-volume-kuberay-workergr-worker-
  labels:
    app.kubernetes.io/created-by: kuberay-operator
    app.kubernetes.io/instance: testcluster-4gib-1gib-2gib-volume
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: kuberay
    helm.sh/chart: ray-cluster-1.2.2
    ray.io/cluster: testcluster-4gib-1gib-2gib-volume-kuberay
    ray.io/group: workergroup
    ray.io/identifier: testcluster-4gib-1gib-2gib-volume-kuberay-worker
    ray.io/is-ray-node: "yes"
    ray.io/node-type: worker
  name: testcluster-4gib-1gib-2gib-volume-kuberay-workergr-worker-h7brl
  namespace: dev-namespace
  ownerReferences:
  - apiVersion: ray.io/v1
    blockOwnerDeletion: true
    controller: true
    kind: RayCluster
    name: testcluster-4gib-1gib-2gib-volume-kuberay
    uid: d71ad0df-2d19-42fb-88c5-f3b1509ba4ad
  resourceVersion: "1805031190"
  uid: 47c1c06f-bcb9-4ae7-aaa0-8e2450ae0f17
spec:
  affinity: {}
  containers:
  - args:
    - 'ulimit -n 65536; ray start  --block  --dashboard-agent-listen-port=52365  --num-cpus=1  --memory=4294967296  --address=testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local:6379  --metrics-export-port=8080 '
    command:
    - /bin/bash
    - -lc
    - --
    env:
    - name: FQ_RAY_IP
      value: testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local
    - name: RAY_IP
      value: testcluster-4gib-1gib-2gib-volume-kuberay-head-svc
    - name: RAY_CLUSTER_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.labels['ray.io/cluster']
    - name: RAY_CLOUD_INSTANCE_ID
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: RAY_NODE_TYPE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.labels['ray.io/group']
    - name: KUBERAY_GEN_RAY_START_CMD
      value: 'ray start  --block  --dashboard-agent-listen-port=52365  --num-cpus=1  --memory=4294967296  --address=testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local:6379  --metrics-export-port=8080 '
    - name: RAY_PORT
      value: "6379"
    - name: RAY_ADDRESS
      value: testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local:6379
    - name: RAY_USAGE_STATS_KUBERAY_IN_USE
      value: "1"
    - name: REDIS_PASSWORD
    - name: RAY_DASHBOARD_ENABLE_K8S_DISK_USAGE
      value: "1"
    image: rayproject/ray:2.40.0-py310
    imagePullPolicy: IfNotPresent
    livenessProbe:
      exec:
        command:
        - bash
        - -c
        - wget -T 2 -q -O- http://localhost:52365/api/local_raylet_healthz | grep
          success
      failureThreshold: 120
      initialDelaySeconds: 30
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 2
    name: ray-worker
    ports:
    - containerPort: 8080
      name: metrics
      protocol: TCP
    readinessProbe:
      exec:
        command:
        - bash
        - -c
        - wget -T 2 -q -O- http://localhost:52365/api/local_raylet_healthz | grep
          success
      failureThreshold: 10
      initialDelaySeconds: 10
      periodSeconds: 5
      successThreshold: 1
      timeoutSeconds: 2
    resources:
      limits:
        cpu: "1"
        memory: 4Gi
      requests:
        cpu: "1"
        ephemeral-storage: 50Mi
        memory: 1G
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      runAsUser: 1000
      seccompProfile:
        type: RuntimeDefault
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /tmp/ray
      name: log-volume
    - mountPath: /dev/shm
      name: shared-mem
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-5qzrx
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  imagePullSecrets:
  - name: harbor
  initContainers:
  - args:
    - "\n\t\t\t\t\tSECONDS=0\n\t\t\t\t\twhile true; do\n\t\t\t\t\t\tif (( SECONDS
      <= 120 )); then\n\t\t\t\t\t\t\tif ray health-check --address testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local:6379
      > /dev/null 2>&1; then\n\t\t\t\t\t\t\t\techo \"GCS is ready.\"\n\t\t\t\t\t\t\t\tbreak\n\t\t\t\t\t\t\tfi\n\t\t\t\t\t\t\techo
      \"$SECONDS seconds elapsed: Waiting for GCS to be ready.\"\n\t\t\t\t\t\telse\n\t\t\t\t\t\t\tif
      ray health-check --address testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local:6379;
      then\n\t\t\t\t\t\t\t\techo \"GCS is ready. Any error messages above can be safely
      ignored.\"\n\t\t\t\t\t\t\t\tbreak\n\t\t\t\t\t\t\tfi\n\t\t\t\t\t\t\techo \"$SECONDS
      seconds elapsed: Still waiting for GCS to be ready. For troubleshooting, refer
      to the FAQ at https://github.com/ray-project/kuberay/blob/master/docs/guidance/FAQ.md.\"\n\t\t\t\t\t\tfi\n\t\t\t\t\t\tsleep
      5\n\t\t\t\t\tdone\n\t\t\t\t"
    command:
    - /bin/bash
    - -lc
    - --
    env:
    - name: FQ_RAY_IP
      value: testcluster-4gib-1gib-2gib-volume-kuberay-head-svc.dev-namespace.svc.cluster.local
    - name: RAY_IP
      value: testcluster-4gib-1gib-2gib-volume-kuberay-head-svc
    image: rayproject/ray:2.40.0-py310
    imagePullPolicy: IfNotPresent
    name: wait-gcs-ready
    resources:
      limits:
        cpu: 200m
        memory: 256Mi
      requests:
        cpu: 200m
        ephemeral-storage: 50Mi
        memory: 256Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      readOnlyRootFilesystem: true
      runAsNonRoot: true
      runAsUser: 1000
      seccompProfile:
        type: RuntimeDefault
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /tmp/ray
      name: log-volume
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-5qzrx
      readOnly: true
  nodeName: [...redacted...]
  preemptionPolicy: Never
  priority: -100
  restartPolicy: Always
  schedulerName: default-scheduler
  securityContext:
    fsGroup: 1000
    fsGroupChangePolicy: OnRootMismatch
  serviceAccount: default
  serviceAccountName: default
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - emptyDir:
      sizeLimit: 2Gi
    name: shared-mem
  - emptyDir: {}
    name: log-volume
  - name: kube-api-access-5qzrx
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
status:
  conditions:
  - lastProbeTime: null
    lastTransitionTime: "2024-12-12T16:04:30Z"
    status: "True"
    type: PodReadyToStartContainers
  - lastProbeTime: null
    lastTransitionTime: "2024-12-12T16:04:56Z"
    status: "True"
    type: Initialized
  - lastProbeTime: null
    lastTransitionTime: "2024-12-12T16:05:09Z"
    status: "True"
    type: Ready
  - lastProbeTime: null
    lastTransitionTime: "2024-12-12T16:05:09Z"
    status: "True"
    type: ContainersReady
  - lastProbeTime: null
    lastTransitionTime: "2024-12-12T16:04:28Z"
    status: "True"
    type: PodScheduled
  containerStatuses:
  - containerID: containerd://7714c6ed2bf8645cfbf593af137b741ab04ae3c05116db86133e5e7a85fb1dff
    image: docker.io/rayproject/ray:2.40.0-py310
    imageID: docker.io/rayproject/ray@sha256:5c2aac392edc9941a8774284ae5d3bf557538477dbe94bf93f3d20c769cd4e60
    lastState: {}
    name: ray-worker
    ready: true
    restartCount: 0
    started: true
    state:
      running:
        startedAt: "2024-12-12T16:04:57Z"
    volumeMounts:
    - mountPath: /tmp/ray
      name: log-volume
    - mountPath: /dev/shm
      name: shared-mem
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-5qzrx
      readOnly: true
      recursiveReadOnly: Disabled
  hostIP: 10.70.3.26
  hostIPs:
  - ip: 10.70.3.26
  initContainerStatuses:
  - containerID: containerd://f9fea6a02f5ff121d504c02f63586b70d65bf0f2e4e328d80fd29422f4f5e5cb
    image: docker.io/rayproject/ray:2.40.0-py310
    imageID: docker.io/rayproject/ray@sha256:5c2aac392edc9941a8774284ae5d3bf557538477dbe94bf93f3d20c769cd4e60
    lastState: {}
    name: wait-gcs-ready
    ready: true
    restartCount: 0
    started: false
    state:
      terminated:
        containerID: containerd://f9fea6a02f5ff121d504c02f63586b70d65bf0f2e4e328d80fd29422f4f5e5cb
        exitCode: 0
        finishedAt: "2024-12-12T16:04:54Z"
        reason: Completed
        startedAt: "2024-12-12T16:04:30Z"
    volumeMounts:
    - mountPath: /tmp/ray
      name: log-volume
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-5qzrx
      readOnly: true
      recursiveReadOnly: Disabled
  phase: Running
  podIP: 10.236.215.168
  podIPs:
  - ip: 10.236.215.168
  qosClass: Burstable
  startTime: "2024-12-12T16:04:28Z"

```

**Comment by andrewsykim** [comment]
I think the problem is that we are checking requests first, and then limit when determining the /dev/shm size? I opened this PR to prioritize limits first, since that's almost always bigger than requests. I'm trying to think if this could be a breaking change for existing deployments though.

**Comment by andrewsykim** [comment]
https://github.com/ray-project/kuberay/pull/2642

**Comment by andrewsykim** [comment]
@jonded94 as a workaround for now, it loosk like you can just match requests / limits for memory

**Comment by andrewsykim** [comment]
> To be honest, I don't really understand why though. How are these things connected?

`/dev/shm` is used for Ray's plasma in-memory store, this is where remote objects are stored. I believe this needs to be some percentage of the available memory for Ray. I don't recall what the minium threshold is, but ~30% sounds familiar. This is likely the reason you're seeing the cluster working only when /dev/shm is > 30% of total memory

**Comment by andrewsykim** [comment]
Ah, here's the 30% reference:

![image](https://github.com/user-attachments/assets/ab376d33-f224-4a23-94cf-5aed5aec41d7)

**Comment by andrewsykim** [comment]
So what's happening is that you set limit to 4GiB, so kuberay starts Ray `ray start --memory=4Gi`, and it then expects 30% of 4Gi to be available within `/dev/shm`. However, the memory request is still 1GiB, so it falls back to `/tmp`, which does not perform well and is likely the reason your cluster is timing out. This behavior would be fixed with https://github.com/ray-project/kuberay/pull/2642, where we adjust the size of `/dev/shm` based on the limit first, then requests

**Comment by jonded94** [comment]
Okay, sounds good! :) Thank you for your extremely quick invesigation. Determining the size of `/dev/shm` through resource limits instead of requests first sounds to be the right way ahead, at least as far as I can see as an outsider.

As a background why I stumbled upon this issue: We have very spiky workloads, but share these big nodes with *a lot* of other infrastructure. So we can't really afford blocking many resources by settings requests == limits. What we would go for right now is just to manually override the `/dev/shm` volume size since local storage is relatively plenty.

**Comment by andrewsykim** [comment]
Manually overriding seems like a  fine workaround

**Comment by andrewsykim** [comment]
We're planning v1.3 this month

**Comment by kevin85421** [comment]
Closed by #2642
